javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+The methods and properties you *do* intend for others to use on your objects might be considered your object's **public interface**, or **public properties**/**public methods**. Having a good, well thought out interface on your objects is important- not only because it makes your object pleasant to use by you and others, but also to keep objects flexible and extensible in the future.
+
+This idea of grouping related functionality within an object is *extremely powerful*, and can often result in more organized, understandable code.
+
+Furthermore, with the various object creation methods you'll learn throughout this section of the curriculum, you'll be able to easily duplicate and reuse objects like these! Imagine you have a website where users can create and play *multiple* rock-paper-scissor games at once. Managing the data and interacting with each of those games would be no sweat with objects!
+
+### Objects As Machines
+
+Sometimes, you may want to create objects that embody complex concepts rather actual, physical items- similar to the RPS game above. Objects can be used to represent almost anything you can think of, and it's impossible to give a comprehensive list of examples. However, a few examples might be:
+
+- An object that manages other objects, such as an "inventory" object that contains a list of items, and actions that can be done with those items.
+- An object that can listen for events that happen, and respond appropriately (think of `.addEventListener` on DOM elements)
+- A "Debt" object that keeps track of a debt owed, how much has been paid, and how much of the debt is remaining.
+
+You may have trouble figuring out what kinds of properties and methods one of these objects might contain at first. One way you might conceptualize these objects might be to imagine them as little 'machines' you're making out of code.
+
+The properties of the machine could be thought of displays that might show information such as:
+
+- A list of the items you've collected, the total amount of items you can carry, and how much you're currently carrying
+- A list of functions that are listening for an event
+- The person who owes a particular debt, and the original amount owed.
+
+The methods of your machine might be akin to buttons and such that make the machinde *do* a specific thing, such as:
+
+- Remove an item you own from a list, add a new item, upgrade an item, craft a new item
+- Fire all the functions that are listening to a 'click' event, or add a new function to listen to the 'click' event
+- Pay an amount of money towards a debt, and determine how much more money is owed on a debt
+
+Again, objects can be used to represent almost anything you can think of, the limit is your imagination!
```

Comment:
I like this section. Two suggestions:

1. How about replacing the "listen for events" example with one that's a little more resemblant of a display controller, like what one will likely use in Tic Tac Toe? I think it might be nice to drop a little early hint here than might also help people connect the dots if they revisit this lesson when doing Tic Tac Toe. "Display controller" often ends up being something that's conceptually tricky to get round.
2. I'd like for this section to make clear that these concepts will become easier to identify and design through experience, including the use of further concepts/techniques introduced in later lessons. For example, the "you may have trouble..." paragraph:
   > You may have trouble figuring out what kinds of properties and methods one of these objects might contain at first. These sorts of things come with experience and concepts taught in later lessons can help as well. One way you might conceptualize these objects might be to imagine them as little 'machines' you're making out of code.

Reply:
> How about replacing the "listen for events" example with one that's a little more resemblant of a display controller, like what one will likely use in Tic Tac Toe? I think it might be nice to drop a little early hint here than might also help people connect the dots if they revisit this lesson when doing Tic Tac Toe. "Display controller" often ends up being something that's conceptually tricky to get round.

I'm assuming you mean something like the ScreenController in Building a House From the Inside out. That seems tricky, and may not perfectly fit in here since in that article it's not something that's implemented as an object, and I think we've been doing a decent job of preventing learners from trying. Also, if they were to try to implement that as an object- I suspect it'd look a lot like the PubSub pattern- which this definitely covers. 

If you meant something more like a "view" in the MVC pattern, I'd agree that a lot of learners struggle to create a simple object that just contains some nice DOM utility functions that perform specific actions- divorced from events. We could add something like that in here, but since a module like that might only contain simple functions, and not particularly complex behavior, it's a weird spot to put it. Not opposed to it though to prime learners to think about it. 

> I'd like for this section to make clear that these concepts will become easier to identify and design through experience, including the use of further concepts/techniques introduced in later lessons. For example, the "you may have trouble..." paragraph:
> > You may have trouble figuring out what kinds of properties and methods one of these objects might contain at first. These sorts of things come with experience and concepts taught in later lessons can help as well. One way you might conceptualize these objects might be to imagine them as little 'machines' you're making out of code.

I like that idea a lot. I'll get it in there.

Reply:
That's fair on the first point - not a vital change of course :)